### PR TITLE
Dev branch

### PR DESCRIPTION
**Pull request by IrinaKostevich**

### CODE REVIEW

File: src/scripts/ui/diagram.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
+const CLASS_ITEM_UP: string = 'diagram-item--up';
+
+export function generateDiagramItemsList(array: number[]): HTMLOListElement {
+    const diagramItemsList: HTMLOListElement = document.createElement('ol');
+    diagramItemsList.classList.add('diagram');
+
+    const items: string[] = array.map((value, index) => {
```
Comment (alexeykostevich) [inline src/scripts/ui/diagram.ts]: Oh my goodness! Fantastic job! 

A `for` loop filling an empty array has finally gone :)

File: src/styles/components/diagram.css [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    margin-right: 0.5rem;
+}
+
+.diagram .diagram-item .item-pipe {
+    width: 25px;
+    background-color: lightblue;
+    margin-bottom: 0.5rem;
+}
+
+.diagram .diagram-item .item-value {
```
Comment (alexeykostevich) [inline src/styles/components/diagram.css]: Should it be `diagram-item-value`?

File: index.html [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+</head>
+
+<body>
+    <header class="header">
+        <p class="header-title">Sorting Visualization</p>
+    </header>
+    <main>
+        <section class="l-diagram-container">
+            <!-- Count Input -->
+            <div class="input">
```
[the remark this answers is not in the document]
Comment (IrinaKostevich) [reply index.html]: Renamed.

File: src/scripts/core/sort-strategies/bubble-sort.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { SortSwapPairIndexes } from '../sort-types.js';
+
+export function* bubbleSort(array: number[]): IterableIterator<SortSwapPairIndexes> {
```
[the remark this answers is not in the document]
Comment (IrinaKostevich) [reply src/scripts/core/sort-strategies/bubble-sort.ts]: Done.

File: src/scripts/core/sort-strategies/bubble-sort.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    const shuffledArray: number[] = [...array];
+    let shuffleNeeded: boolean;
```
[the remark this answers is not in the document]
Comment (IrinaKostevich) [reply src/scripts/core/sort-strategies/bubble-sort.ts]: Done.

File: src/scripts/core/sort-strategies/bubble-sort.ts [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+    for (let j = 0; j < shuffledArray.length; j += 1) {
+        for (let i = 0; i < shuffledArray.length - j - 1; i += 1) {
+            shuffleNeeded = false;
```
[the remark this answers is not in the document]
Comment (IrinaKostevich) [reply src/scripts/core/sort-strategies/bubble-sort.ts]: Done.

File: src/scripts/core/sort-types.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import { bubbleSort } from './sort-strategies/bubble-sort.js';
+import { selectionSort } from './sort-strategies/selection-sort.js';
+
+export type SortValue = 'bubble' | 'selection';
```
[the remark this answers is not in the document]
Comment (IrinaKostevich) [reply src/scripts/core/sort-types.ts]: Done.

File: src/scripts/core/utils.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
+function* generateArrayValue(count: number): IterableIterator<number> {
```
[the remark this answers is not in the document]
Comment (IrinaKostevich) [reply src/scripts/core/utils.ts]: Done.

File: src/scripts/core/utils.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    return shuffledArray;
+}
+
+export function generateArray(length: number): number[] {
+    const array: number[] = [...generateArrayValue(length)];
+
+    return shuffleArray(array);
+}
+
+export function timer(ms: number) {
```
[the remark this answers is not in the document]
Comment (IrinaKostevich) [reply src/scripts/core/utils.ts]: Done.

File: src/scripts/ui/app.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+
+    onNumberChange(event: Event) {
+        const arrayItemsCount: number = (<HTMLInputElement>event.target).valueAsNumber;
+
+        if (arrayItemsCount < MIN_COUNT) return;
+        if (arrayItemsCount > MAX_COUNT) return;
+
+        this.array = generateArray(arrayItemsCount);
+
+        const diagramItemsList: HTMLOListElement = generateDiagramItemsList(this.array);
```
[the remark this answers is not in the document]
Comment (IrinaKostevich) [reply src/scripts/ui/app.ts]: `generateDiagramItemsList(this.array);` returns an element <ol>, so I thought it's better be to be more specific in HTML element type.

File: src/scripts/ui/app.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        const diagramChart: HTMLElement = ELEMENTS.diagramChart(this.document);
+
+        this.renderElementWithChildren(diagramChart, [diagramItemsList]);
+    }
+
+    onSortClick() {
+        const sortValue: SortValue = ELEMENTS.sortDropdown(this.document).value as SortValue;
+
+        if (!sortValue) {
```
[the remark this answers is not in the document]
Comment (IrinaKostevich) [reply src/scripts/ui/app.ts]: Renamed.

File: src/scripts/ui/app.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+        const sortFunction: SortFunction = sortMap.get(sortValue) as SortFunction;
+
+        for (const { index1, index2, shuffleNeeded } of sortFunction(array)) {
+            const item1: HTMLElement = this.document.querySelector(`li[data-order='${index1}']`) as HTMLElement;
+            const item2: HTMLElement = this.document.querySelector(`li[data-order='${index2}']`) as HTMLElement;
+    
+            showComparingItems(item1, item2);
+    
+            await timer(1000);
+            if (shuffleNeeded) swithchTwoDiagramItems(item1, item2);
```
[the remark this answers is not in the document]
Comment (IrinaKostevich) [reply src/scripts/ui/app.ts]: Renamed.

File: src/scripts/ui/app.ts [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    
+            await timer(1000);
+            hideComparingItems(item1, item2);
+        }
+    }
+
+    renderElementWithChildren(element: HTMLElement, children: HTMLElement[]) {
```
[the remark this answers is not in the document]
Comment (IrinaKostevich) [reply src/scripts/ui/app.ts]: Renamed.

[the remark this answers is not in the document]
Comment (IrinaKostevich) [reply src/scripts/ui/app.ts]: Should I also update the return type for class methods as well?

File: src/styles/components/diagram.css (hunk 2 above)
Comment (IrinaKostevich) [reply src/styles/components/diagram.css]: Renamed.

File: src/scripts/ui/app.ts [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        const sortButton: HTMLElement = ELEMENTS.sortButton(this.document);
+
+        numberInput.addEventListener('change', this.onNumberChange.bind(this));
+        sortButton.addEventListener('click', this.onSortClick.bind(this));
+    }
+
+    onNumberChange(event: Event) {
+        const arrayItemsCount: number = (<HTMLInputElement>event.target).valueAsNumber;
+
+        if (arrayItemsCount < MIN_COUNT) return;
```
[the remark this answers is not in the document]
Comment (IrinaKostevich) [reply src/scripts/ui/app.ts]: Done with form validation.

File: src/scripts/core/sort-strategies/selection-sort.ts [hunk 15]
```diff
@@ -0,0 +1,23 @@
+import { SortSwapPairIndexes, SortingFn } from '../sort-types.js';
+
+export const selectionSort: SortingFn = function* (array: number[]): IterableIterator<SortSwapPairIndexes> {
+    const sortedArray: number[] = [...array];
+
+    for (let j = 0; j < sortedArray.length - 1; j += 1) {
+        let minIndex = j;
```
Comment (alexeykostevich) [inline src/scripts/core/sort-strategies/selection-sort.ts]: Please, aspire to specify a type for all the variables.

File: src/scripts/ui/app.ts [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+    diagramChart: (document: Document) => document.querySelector('.js-diagram-chart') as HTMLElement,
+    form: (document: Document) => document.querySelector('.js-form') as HTMLFormElement
+};
+
+export class App {
+    private _array: number[] = [];
+
+    constructor(private document: Document) { }
+
+    set array(value: number[]) {
```
Comment (alexeykostevich) [inline src/scripts/ui/app.ts]: Please, add a return type.

```ts
set array(value: number[]): void {
}
```

File: src/scripts/ui/app.ts [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+        this._array = value;
+    }
+
+    get array() {
```
Comment (alexeykostevich) [inline src/scripts/ui/app.ts]: Please, add a return type.

```ts
get array(): number[] {
}
```

File: src/scripts/ui/app.ts [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+        return this._array;
+    }
+
+    start() {
```
Comment (alexeykostevich) [inline src/scripts/ui/app.ts]: Please, add a return type here.

```ts
start(): void {
}
```

The same issue exists in all the methods below :)

File: src/scripts/core/utils.ts [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+    return shuffledArray;
+}
+
+export function generateArray(length: number): number[] {
+    const array: number[] = [...range(1, length)];
+
+    return shuffleArray(array);
+}
+
+export function wait(ms: number) {
```
Comment (alexeykostevich) [inline src/scripts/core/utils.ts]: You could try to use this declaration, which means a *Promise* that resolves with no value.

```ts
export function wait(ms: number): Promise<void> {
}
```

If the TS compiler does not see the type, take a look at this:
https://mariusschulz.com/blog/built-in-type-declarations-in-typescript

File: src/scripts/core/sort-strategies/selection-sort.ts (hunk 15 above)
Comment (IrinaKostevich) [reply src/scripts/core/sort-strategies/selection-sort.ts]: Done.

File: src/scripts/ui/app.ts (hunk 17 above)
Comment (IrinaKostevich) [reply src/scripts/ui/app.ts]: Done.

File: src/scripts/ui/app.ts (hunk 16 above)
Comment (IrinaKostevich) [reply src/scripts/ui/app.ts]: It's not allowed to set return type for set.

File: src/scripts/ui/app.ts (hunk 18 above)
Comment (IrinaKostevich) [reply src/scripts/ui/app.ts]: Done.

File: src/scripts/ui/app.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (IrinaKostevich) [reply src/scripts/ui/app.ts]: Done.

File: src/scripts/core/utils.ts (hunk 19 above)
Comment (IrinaKostevich) [reply src/scripts/core/utils.ts]: Done.

**Pull request merged**